packages/headless/src/controllers/folding/headless-folding.ts

```diff
@@ -0,0 +1,125 @@
+import {NumberValue, Schema, StringValue} from '@coveo/bueno';
+import {Engine} from '../../app/headless-engine';
+import {search, configuration, folding} from '../../app/reducers';
+import {registerFolding} from '../../features/folding/folding-actions';
+import {FoldedResult} from '../../features/folding/folding-state';
+import {
+  ConfigurationSection,
+  FoldingSection,
+  SearchSection,
+} from '../../state/state-sections';
+import {loadReducerError} from '../../utils/errors';
+import {validateOptions} from '../../utils/validate-payload';
+import {buildController, Controller} from '../controller/headless-controller';
+
+export {FoldedResult};
+
+const optionsSchema = new Schema<Required<FoldingOptions>>({
+  collectionField: new StringValue(),
+  parentField: new StringValue(),
+  childField: new StringValue(),
+  numberOfFoldedResults: new NumberValue(),
+});
+
+export interface FoldingOptions {
+  /**
+   * The name of the field used to resolve what collection a result is part of.
+   *
+   * @defaultValue `foldedcollection`
+   */
+  collectionField?: string;
+  /**
+   * The name of the field used to reference which result is the parent of another one within a collection.
+   *
+   * @defaultValue `foldedparent`
+   */
+  parentField?: string;
+  /**
+   * The name of the field used to uniquely identify a result within a collection.
+   *
+   * @defaultValue `foldedchild`
+   */
+  childField?: string;
+  /**
+   * The number of additional results to fetch from the same collection as
+   * the most relevant result, excluding the root result of the collection.
+   *
+   * @defaultValue `2`
+   */
+  numberOfFoldedResults?: number;
+}
+
+export interface FoldingProps {
+  /**
+   * The options for the `Folding` controller.
+   * */
+  options?: FoldingOptions;
+}
+
+/**
+ * The `Folding` headless controller re-organizes results into hierarchical collections (a.k.a. threads).
+ */
+export interface Folding extends Controller {
+  /**
+   * The state of the `Folding` controller.
+   */
+  state: FoldingState;
+}
+
+/**
+ * A scoped and simplified part of the headless state that is relevant to the `Folding` controller.
+ * */
+export interface FoldingState {
+  /**
+   * The unsorted hierarchical collections of results.
+   * */
+  collections: FoldedResult[];
+}
+
+/**
+ * Creates a `Folding` controller instance.
+ *
+ * @param engine - The headless engine.
+ * @param props - The configurable `Folding` properties.
+ * @returns A `Folding` controller instance.
+ */
+export function buildFolding(
```

Comment:
One thing we did not catch in the design doc, and that is now more apparent with this PR (especially the react sample you built), is the fact that it's actually kind of a "replacement" for the current headless-result-list controller...

As in, you would either use the standard headless result list if you are not doing any folding request, and the folding result list otherwise.

So, it seems to me we need to get back a bit on the drawing board: Either we extend the current headless-result-list to support optional folding parameters, or we embed `headless-result-list` in this controller (composition) with the added feature needed for folding.

Because, for example, as a end user that needs to do a folded result list, they would still want to be able to specify the `fieldsToInclude` and to be able to `fetchMoreResults()`.

Reply:
I think I would personally vote for embedding `headless-result-list` in here.

Now... I'm pretty sure the documentation is going to blow up ? :)

Reply:
I agree, I'm not 100% sure yet whether I prefer folding to be a separate controller or part of result list, but extending result list sounds like a good idea to me.

I think the doc would handle it just fine since we support `extend`.

Reply:
If we embed the result list in there, I think the name should be changed to reflect that.

```
FoldedResultList
FoldedResultListState
buildFoldedResultList
etc.
```